community/kernel/src/test/java/org/neo4j/kernel/impl/locking/community/RWLockTest.java

```diff
@@ -0,0 +1,456 @@
+
+        // when
+        lock.terminateLockRequestsForLockTransaction( readerTransaction );
+        lock.terminateLockRequestsForLockTransaction( writerTransaction );
+
+        readerCompletedLatch.await();
+        writerCompletedLatch.await();
+
+
+        // expect only main write lock counters and elements present
+        // all the rest should be cleaned up
+        Assert.assertEquals( 0, lock.getWaitingThreadsCount() );
+        Assert.assertEquals( 0, lock.getReadCount() );
+        Assert.assertEquals( 1, lock.getWriteCount() );
+        Assert.assertEquals( 1, lock.getTxLockElementCount() );
+
+    }
+
+    private Runnable createReader( final RWLock lock, final LockTransaction transaction,
+            final CountDownLatch latch )
+    {
+        return new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                lock.acquireReadLock( transaction );
+                latch.countDown();
+            }
+        };
+    }
+
+    private Runnable createFailedReader( final RWLock lock, final LockTransaction transaction,
+            final CountDownLatch latch )
+    {
+        return new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                Assert.assertFalse( lock.acquireReadLock( transaction ) );
+                latch.countDown();
+            }
+        };
+    }
+
+    private Runnable createWriter( final RWLock lock, final LockTransaction transaction,
+            final CountDownLatch latch )
+    {
+        return new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                lock.acquireWriteLock( transaction );
+                latch.countDown();
+            }
+        };
+    }
+
+    private Runnable createFailedWriter( final RWLock lock, final LockTransaction transaction,
+            final CountDownLatch latch )
+    {
+        return new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                Assert.assertFalse( lock.acquireWriteLock( transaction ) );
+                latch.countDown();
+            }
+        };
+    }
+
+    private Runnable createReaderForDeadlock( final RWLock node, final LockTransaction transaction,
+            final CountDownLatch latch )
+    {
+        return new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    node.acquireReadLock( transaction );
+                }
+                catch ( DeadlockDetectedException e )
+                {
+                    latch.countDown();
+                }
+            }
+        };
+    }
+
+    private void waitWaitingThreads( RWLock lock, int expectedThreads ) throws InterruptedException
+    {
+        while ( lock.getWaitingThreadsCount() != expectedThreads )
+        {
+            Thread.sleep( 100 );
```

Comment:
I would guess that reducing the wait time here, could speed up the test a little bit.